### PR TITLE
SBLK_MASK_LOCK_THREADID allow tid up to 65535

### LINKED ISSUE #88426: Long Pause on Gen1 GC
### Description

Recently my team has been working on lowering down memory allocation to minimize GC pressure.
While we used different technique to reuse/recycle object, or use ArrayPool.Shared to rent/return memory, we found that there long periodic long pauses (~1s) of Gen0/Gen1 GC few times per hour.
We used PerfView command to spot it out:
```
PerfView -NoGui -AcceptEULA -NoNGenRundown -BufferSizeMB:2048 -CircularMB:2048 -Zip:false -Merge:true -ClrEvents:GC+Stack -StopOnGCOverMSec:300 -KernelEvents:Memory,VirtualAlloc,Default -OnlyProviders:ClrPrivate:1:5,Clr:0x40000001:5 -NoRundown
-DelayAfterTriggerSec:0 -CollectMultiple:1 -DataFile:XXX-2307051821-48272-gcsg -FocusProcess:48272 -Process:48272 collect
```

GC Stats showing 1s Gen1 GC pause
Screenshot 2023-07-05 225019
[pic1]

When looking into the CPU Stack of profile, we discovered that these few calls during GC causes most part of delay:
1. coreclr!SyncBlockCache::InsertCleanupSyncBlock
2. coreclr!SVR::GCHeap::IsPromoted
3. ntoskrnl!MiRemoveLowestPriorityStandbyPage
Screenshot 2023-07-05 224639
[pic2]

When Goto Source these lines are most time consuming ...
InsertCleanupSyncBlock
Screenshot 2023-07-05 231740
[pic3]

IsPromoted
Screenshot 2023-07-05 231840
[pic4]

### Configuration
.NET core 7
Windows Server 2012 x64

### PR DESCRIPTION
**Pull request by dickens-code**
This is an attempt to fix https://github.com/dotnet/runtime/issues/88426 by avoiding sync block creation.

All test failures are known and is unrelated to the change.

@dotnet/gc

### CODE REVIEW

**Comment by cshung** [comment]
@jkotas, can you please take a look and see if there will be problems if we increase the number of bits used for thread id?

File: src/coreclr/vm/syncblk.h [hunk 1]
```diff
@@ -103,10 +103,10 @@ typedef DPTR(EnCSyncBlockInfo) PTR_EnCSyncBlockInfo;
 //   value is zero if no thread is holding the lock
 // - following six bits (bits 10 thru 15) is recursion level used for the thin locks
 //   value is zero if lock is not taken or only taken once by the same thread
-#define SBLK_MASK_LOCK_THREADID             0x000003FF   // special value of 0 + 1023 thread ids
-#define SBLK_MASK_LOCK_RECLEVEL             0x0000FC00   // 64 recursion levels
-#define SBLK_LOCK_RECLEVEL_INC              0x00000400   // each level is this much higher than the previous one
-#define SBLK_RECLEVEL_SHIFT                 10           // shift right this much to get recursion level
+#define SBLK_MASK_LOCK_THREADID             0x0000FFFF   // special value of 0 + 65535 thread ids
```
Comment (jkotas) [inline src/coreclr/vm/syncblk.h:106]: > lower ten bits (bits 0 thru 9) is thread id used for the thin locks

This comment needs updating as well.

**Comment by jkotas** [comment]
> @jkotas, can you please take a look and see if there will be problems if we increase the number of bits used for thread id?

I do not see a problem with it. cc @VSadov @kouvel

**Comment by jkotas** [comment]
> This is an attempt to fix https://github.com/dotnet/runtime/issues/88426 by avoiding sync block creation.

How many threads does the workload in question have? This change is only going to help if it has more than 1000 threads. If the syncblocks are created for other reasons, it is not going to help.

**Comment by VSadov** [comment]
> How many threads does the workload in question have? 

Note that this is not only about continuous number of threads, but about the highest thread ID that threads can have. 

Thread pool often runs number of threads more than number of cores. And a high-end machine may have hundreds of logical cores this days, so 1024+ threads seems possible.

But also, there is some latency in thread termination and ID recycling. It may be possible for some threads after a burst of activity to get IDs higher than 1024 even if you did not have this many at the same time, so you may not need a very "high-end" machine, to see this. Once you get such threads in the process, they will inflate every thin lock that they touch.

65535 may be enough for some time, so this is a reasonable change.

**Comment by VSadov** [comment]
For the syncblocks to become a problem you'd need to lock on a large number of objects. 
It is possible in patterns that use fine-grained locking - like some UI framework may lock on widgets, (WebKit does that, but I do not know any examples in .Net). Some graph algorithms could conceivably lock nodes while updating. 

There are other reasons for syncblocks, though - like COM interop. 
While this change makes sense on its own, it may be worth looking at what causes syncblocks to be created.

**Comment by Maoni0** [comment]
the reason why Andrew asked the question is because while this is not a problem on its own, we wanted to make sure we are not robbing other things of using these bits, eg, if this means something else will have 6 fewer bits to use, that's not good. but if when the bits are being used for locking, nothing else will share these 32 bit anyway (ie, those 6 bits wouldn't be used anyway), then it doesn't affect anything else. if anyone has the answer to that question that'd be really helpful. last time I looked at the code for who uses these bits was ages ago so I don't remember the details.

**Comment by VSadov** [comment]
On CoreCLR the sync word has 11 bits that are not used for anything right now. 
This change will leave 5 free bits for the future use - like if we find 65535 too small to store thread IDs in a few years, for example.

**Comment by VSadov** [comment]
> On CoreCLR the sync word has 10 bits that are not used for anything right now.

Well, that is not entirely correct. That is only true when object header is used for locking.
When hashcode is stored in the header it uses 26 bits and leaves no free ones.

**Comment by Maoni0** [comment]
thanks for the info, @VSadov!

we are doing some investigation on our side to see what exactly the culprit of the perf problem was, so please do not merge this for the time being while we are doing that.

**Comment by Maoni0** [comment]
@cshung and I talked offline and concluded that this is a reasonable fix.

File: src/coreclr/vm/syncblk.h (hunk 1 above)
Comment (dickens-code) [reply src/coreclr/vm/syncblk.h:106]: comment updated

**Review by VSadov — APPROVED** [review]
LGTM. Thanks!

**Comment by dickens-code** [comment]
thanks very much @VSadov @Maoni0 @cshung @jkotas :)

**Pull request merged**